core/src/main/kotlin/com/tschuchort/compiletesting/KotlinJsCompilation.kt

```diff
@@ -0,0 +1,392 @@
+
+  /** Disables pre-IR backend */
+  var irOnly: Boolean = false
+
+  /** Specify a compilation module name for IR backend */
+  var irModuleName: String? = null
+
+
+  /** Additional string arguments to the Kotlin compiler */
+  var kotlincArguments: List<String> = emptyList()
+
+  /** Options to be passed to compiler plugins: -P plugin:<pluginId>:<optionName>=<value>*/
+  var pluginOptions: List<PluginOption> = emptyList()
+
+  /**
+   * Path to the kotlin-stdlib-common.jar
+   * If none is given, it will be searched for in the host
+   * process' classpaths
+   */
+  var kotlinStdLibCommonJar: File? by default {
+    findInHostClasspath(hostClasspaths, "kotlin-stdlib-common.jar",
+      kotlinDependencyRegex("kotlin-stdlib-common"))
+  }
+
+  /**
+   * Path to the kotlin-stdlib-js.jar
+   * If none is given, it will be searched for in the host
+   * process' classpaths
+   */
+  var kotlinStdLibJsJar: File? by default {
+    findInHostClasspath(hostClasspaths, "kotlin-stdlib-js.jar",
+      kotlinDependencyRegex("kotlin-stdlib-js"))
+  }
+
+  // Directory for input source files
+  private val sourcesDir get() = workingDir.resolve("sources")
+
+  // *.class files, Jars and resources (non-temporary) that are created by the
+  // compilation will land here
+  val outputDir get() = workingDir.resolve("output")
+
+  /** Result of the compilation */
+  inner class Result(
+    /** The exit code of the compilation */
+    val exitCode: ExitCode,
+    /** Messages that were printed by the compilation */
+    val messages: String
+  ) {
+    /** class loader to load the compile classes */
+    val classLoader = URLClassLoader(arrayOf(outputDirectory.toURI().toURL()),
+      this::class.java.classLoader)
+
+    /** The directory where only the final output class and resources files will be */
+    val outputDirectory: File get() = outputDir
+
+    /**
+     * Compiled class and resource files that are the final result of the compilation.
+     */
+    val compiledClassAndResourceFiles: List<File> = outputDirectory.listFilesRecursively()
+  }
+
+
+  // setup common arguments for the two kotlinc calls
+  private fun commonK2JSArgs() = K2JSCompilerArguments().also { args ->
+    args.pluginClasspaths = pluginClasspaths.map(File::getAbsolutePath).toTypedArray()
+
+    args.verbose = verbose
+
+    // the compiler should never look for stdlib or reflect in the
+    // kotlinHome directory (which is null anyway). We will put them
+    // in the classpath manually if they're needed
+    args.noStdlib = true
+
+    args.sourceMap = true
+    args.listPhases = true
+    args.moduleKind = "commonjs"
+    args.outputFile = File(outputDir, outputFileName).absolutePath
+    args.metaInfo = true
+    args.sourceMapBaseDirs = commonClasspaths().joinToString(separator = File.pathSeparator)
+    args.libraries = listOfNotNull(kotlinStdLibJsJar).joinToString(separator = ":")
+
+    args.irProduceKlibDir = irProduceKlibDir
```

Comment:
These options are unique to the JS compiler. JVM compiler only has a `useIR` option. If we want to limit the number of options initially the important one is `irProduceJs` for testing IR based compiler plugins. This also brings up the question if the library should abstract these options or use the same naming as the underlying options.

Reply:
I don't want to limit the available options, I was just wondering if they should also be added to the Java KotlinCompilation. I usually try to use the same naming as in the `K2JVMCompilerArguments` class. Originally I did it this way instead of letting the user pass in an instance of `K2JVMCompilerArguments` because I wanted to hide all the internals and dependency on kotlin-compiler-embeddable from the user, but I think this is no longer possible because kotlin-compiler-embeddable also declares the `ComponentRegistrar` interface, which users need for compiler plugins. Giving the user full access to `K2JVMCompilerArguments` is probably not a good idea since I am changing the options and it could be confusing.